platform/measurement_utils.cpp

```diff
@@ -23,12 +23,31 @@ using namespace settings;
 using namespace std;
 using namespace strings;
 
+// Global Locale pointer to store system locale.
+Locale *pSysLocale = NULL;
+
+void refreshSystemLocale()
+{
+  if (pSysLocale != NULL)
+    delete pSysLocale;
+
+  pSysLocale = new Locale(GetCurrentLocale());
+}
+
+void setSystemLocale(platform::Locale newLocale)
+{
+  if (pSysLocale != NULL)
+    delete pSysLocale;
+
+  pSysLocale = new Locale(newLocale);
+}
+
 string ToStringPrecision(double d, int pr)
 {
-  // We assume that the app will be restarted if a user changes device's locale.
-  static Locale const loc = GetCurrentLocale();
+  if (pSysLocale == NULL)
+    pSysLocale = new Locale(GetCurrentLocale());
```

Comment:
```suggestion
  SetSystemLocale(GetCurrentLocale());
```

Reply:
@biodranik: Please be aware that, in navigation mode, this `ToStringPrecisionLocale()` function is called 60 times per second: 20 Hz due to compass update and x3 values to format (total distance, next turn distance and speed), so we shall not make calls to `GetCurrentLocale()` this often.

That's why I was using a pointer to store the system locale. It's first set to null and it's only initialized once. The only overhead that we were adding is the check if the pointer is null.

If you don't like using pointers, then at least we shall use an additional static boolean variable to make a only-one-time call to `GetCurrentLocale()`, adding a minimum overhead:

```
static boolean firstTime = true;

if (firstTime)
{
  SetSystemLocale(GetCurrentLocale());

  firstTime = false;
}
```
I will amend the commit using this last approach.

Reply:
OMG that is an issue then. There is no need to format strings so often. Any ideas on how to avoid calling it 60 times per second?

Reply:
I'm starting to think that this code has become too overcomplicated. Let's think about how can we simplify it and avoid unnecessary overhead. Note that calling JNI 60 times per second is very expensive.

Also calling C++ to call Java from it to get locale to init in back in C++ looks very weird and complex. A cleaner solution would just set it once when necessary from Java, if we stick to this approach.

Can we simplify this whole code by implementing it natively in Java and iOS? Any other ideas?

Reply:
We already did the trade-off analysis of the different possible implementations for the locale distance/speed formatting in PR https://github.com/organicmaps/organicmaps/pull/4029#issuecomment-1346347205, and we came with the solution of implementing the formatting in the c++ side, due to these main reasons:

* Faster formatting in c++ than in Java/iOS side.
* Unique place for formatting implementation, otherwise we would have to make implementations for every platform (Android, iOS, macOS, Linux, Windows).

Regarding the 60 times per second update, before this PR we were calling the `GetCurrentLocale()` function only once (at app startup) as we were using a local static variable inside of the `ToStringPrecisionLocale()` function.

With the amend that I've just made 1 hour ago, a boolean variable just prevents from calling this functions more than once.

From my POV, the code is OK as it is now. The only remaining issue in this PR is to check if we can also detect system locale changes in iOS. I will investigate it.